.github/workflows/markdown-lint.yml

```diff
@@ -23,34 +23,59 @@ jobs:
-          ERR_MSG_LINE="::error ${ERR_DETAILS}::ERROR ${ERR_MSG_STUB}"
-          yamllint --version || { printf "%s\n" "${ERR_MSG_LINE}"; exit 1; }
+          sudo apt-get update || exit 1
+          sudo apt-get install -y yamllint npm || :  # handle error with detail below
+          print_error() {
+            local msg="$1"
+            local file_stub=".github/workflows/markdown-lint.yml"
+            local line_start="$2"
+            local line_end="$3"
+            printf "::error file=%s,line=%s,endLine=%s,title=VALIDATION_ERROR::ERROR %s\n" \
+              "${file_stub}" "${line_start}" "${line_end}" "${msg}"
+          }
+          command -v npm >/dev/null || { print_error "npm installation failed." 26 27 ; exit 126; }
+          yamllint --version || { print_error "Yamllint installation failed." 26 27 ; exit 126; }
 
       - name: Lint Workflow YAML
+        if: ${{ success() }}
         run: |
-          yamllint ${{ vars.YAML_ARGS }} .github/workflows/markdown-lint.yml
-
-      - name: Install NPM Dependencies
-        run: |
-          sudo npm install -g markdownlint
-          sudo npm install -g markdownlint-cli
+          YAML_ARGS="${{ vars.YAML_ARGS }}"
+          if [[ "${YAML_ARGS}" =~ ^*[!a-zA-Z0-9\ \-\.]*$ ]]; then
+              ERR_MSG_STUB="Invalid YAML_ARGS: contains unauthorized characters!"
+              ERR_LOC="file=.github/workflows/markdown-lint.yml,line=42,endLine=43,"
+              printf "::error %s%s\n" "${ERR_LOC}" "title=VALIDATION_ERROR::ERROR ${ERR_MSG_STUB}"
+              exit 1
+          else
+              printf "::notice file=.github/workflows/markdown-lint.yml,%s %s\n" \
+                "line=53,endLine=53,title=Validation::Validating workflow with args:" \
+                "${YAML_ARGS}"
+              read -ra ARGS <<< "${YAML_ARGS}"
+              yamllint "${ARGS[@]}" .github/workflows/markdown-lint.yml
+          fi
 
       - name: Lint YAML config for markdown
+        if: ${{ success() }}
         run: |
-          yamllint -f github --no-warnings .markdownlint.yaml
+          YAML_ARGS="${{ vars.YAML_ARGS }}"
+          read -ra ARGS <<< "${YAML_ARGS}"
+          yamllint "${ARGS[@]}" .markdownlint.yaml
 
-      - name: Lint Markdown Files with YAML config
-        if: ${{ !cancelled() }}
+      - name: Install NPM Dependencies
+        if: ${{ success() }}
         run: |
-          git ls-files --exclude-standard -z -- *.md **/*.md |\
-          xargs -0 -I{} markdownlint "{}" --config .markdownlint.yaml ;
+          # npm install
+          ERR_MSG="NPM package installation failed"
+          ERR_LOC_1="file=.github/workflows/markdown-lint.yml,line=70,endLine=70"
+          ERR_LOC_2="file=package.json,line=2,endLine=4"
+          if ! NPM_ERROR=$(npm install 2>&1); then
+          for LOC in "${ERR_LOC_1}" "${ERR_LOC_2}"; do
+            printf "::error %s,title=INSTALL_ERROR::ERROR %s: %s\n" \
+              "${LOC}" "${ERR_MSG}" "${NPM_ERROR}"
+          done
+          fi
```

Comment:
https://github.com/coderabbitai, This change almost seems like it was overkill now, it caused more churn to debug than expected and does not feel better overall. Let's re-analyze (as a post-mortem) the changes to "Install NPM Dependencies" step in .github/workflows/markdown-lint.yml, considering the impacts to metrics like: Code-Size, Maintainability, Security, Stability, Complexity, Readability, in addition to other relevant metrics, (of-course, as a post mortem, include both positive and negative impacts of each metric). Thanks in advance.

Reply:
@coderabbitai?